lib/resty/rediscluster.lua

```diff
@@ -171,30 +192,68 @@ local function try_hosts_slots(self, serv_list)
                     local sub_info = slots_info[n]
-                    --slot info item 1 and 2 are the subrange start end slots
+                    -- slot info item 1 and 2 are the subrange start end slots
                     local start_slot, end_slot = sub_info[1], sub_info[2]
 
-                    -- generate new list of servers
+                    -- generate new list of servers, including IPs and hostnames
+                    local sub_info_j
+                    local sub_info_j_ip
                     for j = 3, #sub_info do
-                      servers.serv_list[#servers.serv_list + 1 ] = { ip = sub_info[j][1], port = sub_info[j][2] }
+                        -- Redis 6: { "172.26.0.21", 6379, "b050a28d860201c487891ecde77efc8a3935ea51" }
+                        -- Redis 7: { "rc-node-1", 6379, "3a4f2814802ca25a5c92dfc167674802126557d3", { "ip", "172.26.0.11" } }
+                        -- Redis 7: { "172.26.0.11", 6379, "64ff3bc4e8f4587c92bcfaa6c085e67417d44d5d", {} }
+                        sub_info_j = sub_info[j]
+                        -- Redis 6: nil
+                        -- Redis 7: { "ip", "172.26.0.11" }
+                        -- Redis 7: {}
+                        sub_info_j_ip = sub_info_j[4]
+                        if sub_info_j_ip and type(sub_info_j_ip) == "table" and
+                           next(sub_info_j_ip) and sub_info_j_ip[2] ~= sub_info_j[1]
+                        then
+                            redis7_with_host = true
+                            -- prioritize IP address over hostname as IP does not suffer from stale DNS cache
+                            servers.serv_list[#servers.serv_list + 1] = { ip = sub_info_j_ip[2], port = sub_info_j[2] }
+                        else
+                            redis7_with_host = false
+                        end
+                        servers.serv_list[#servers.serv_list + 1] = { ip = sub_info_j[1], port = sub_info_j[2] }
                     end
 
                     for slot = start_slot, end_slot do
+                        -- todo: there is no need to create the 'list', as all the slots
+                        -- within [start_slot, end_slot] share the same 'servers'.
+                        -- just do: slots[slot] = servers
                         local list = { serv_list = {} }
-                        --from 3, here lists the host/port/nodeid of in charge nodes
+                        -- from 3, here lists the host/port/nodeid of in charge nodes
                         for j = 3, #sub_info do
-                            list.serv_list[#list.serv_list + 1] = { ip = sub_info[j][1], port = sub_info[j][2] }
+                            sub_info_j = sub_info[j]
+                            sub_info_j_ip = sub_info_j[4]
+                            if sub_info_j_ip and type(sub_info_j_ip) == "table" and
```

Comment:
The code here introduces high coupling throughout the module. Now in many places you need to know that hostnames are first and then there are ips as intertwined pairs - which can be seen in the need of similar comments throught the code: _"-- prioritize IP address over hostname as IP does not suffer from stale DNS cache"_. 
It can also be seen as ripple effects in `pick_node` function that accepts `ip_or_host` argument (which is a [flag argument](https://martinfowler.com/bliki/FlagArgument.html)) and performs custom logic on how this table right here is structured.
Changing this structure in future would require a change in a lot of other places. Therefore I suggest we encapsulate it in an object and hide it there.

Reply:
Will refactor this into a new function as mentioned above. Changing the `ip_or_host` to `ip_host` might be better.

Reply:
done